# Chapter 2: Variables, Expressions, and Statements

## Section: Variables

### 💪 The Power of Variables

One of the **most powerful features** of a programming language is the ability to manipulate **variables**.

#### What is a Variable?

A **variable** is a **name** that refers to a **value**.

Think of it like a labeled box:
- The **label** (variable name) helps you find the box
- The **contents** (value) is what's stored inside
- You can change the contents anytime!

```
┌─────────────┐
│   message   │  ← Variable name (the label)
├─────────────┤
│ "Hello!"    │  ← Value (the contents)
└─────────────┘
```

### 📝 Assignment Statements

An **assignment statement** creates new variables and gives them values.

**Syntax:**
```python
variable_name = value
```

**Key Points:**
- The `=` sign means "assign" (not "equals"!)
- Variable name goes on the **left**
- Value goes on the **right**
- Read it as: "variable_name gets the value"

### 🎯 Examples from the Book

Let's recreate the three assignment examples:

In [1]:
# First assignment: assign a string to variable 'message'
message = 'And now for something completely different'

# Second assignment: assign integer 17 to variable 'n'
n = 17

# Third assignment: assign the approximate value of π to variable 'pi'
pi = 3.1415926535897931

# Let's see what we created
print("Message:", message)
print("N:", n)
print("Pi:", pi)

Message: And now for something completely different
N: 17
Pi: 3.141592653589793


### 🔍 Understanding Each Assignment

Let's break down what happened in each assignment:

#### Assignment 1: String Variable
```python
message = 'And now for something completely different'
```

- **Variable name:** `message`
- **Value type:** String (text)
- **What happened:** Created a new variable called `message` and stored a string in it

In [2]:
# Check the type of message
print(type(message))
print(message)

<class 'str'>
And now for something completely different


#### Assignment 2: Integer Variable
```python
n = 17
```

- **Variable name:** `n`
- **Value type:** Integer (whole number)
- **What happened:** Created a new variable called `n` and stored the number 17 in it

In [3]:
# Check the type of n
print(type(n))
print(n)

<class 'int'>
17


#### Assignment 3: Float Variable
```python
pi = 3.1415926535897931
```

- **Variable name:** `pi`
- **Value type:** Float (decimal number)
- **What happened:** Created a new variable called `pi` and stored the approximate value of π in it

**New type discovered!** This is a **float** (floating-point number) - a number with a decimal point.

In [4]:
# Check the type of pi
print(type(pi))
print(pi)

<class 'float'>
3.141592653589793


### 🔄 Variables Can Change

One powerful feature: you can **reassign** variables to new values!

In [5]:
# Start with a value
x = 5
print("First value:", x)

# Change it to a new value
x = 10
print("New value:", x)

# You can even change the type!
x = "Now I'm a string!"
print("Changed to string:", x)

First value: 5
New value: 10
Changed to string: Now I'm a string!


### 📋 Summary of Data Types

We've now seen **three** main data types:

| Type | Python Name | Example | Description |
|------|-------------|---------|-------------|
| **String** | `str` | `'Hello'` | Text in quotes |
| **Integer** | `int` | `17` | Whole numbers |
| **Float** | `float` | `3.14` | Decimal numbers |

---
### 📝 Practice Exercises

Now it's your turn to create variables!

In [6]:
# Exercise 1: Create a variable called 'name' and assign your name to it
# Then print it
# Your code here:
name = "Alice"
print(name)

Alice


In [7]:
# Exercise 2: Create a variable called 'age' and assign your age to it
# Check its type using type()
# Your code here:
age = 30
print(type(age))

<class 'int'>


In [ ]:
# Exercise 3: Create a variable called 'height' with your height in meters (use decimal)
# Example: 1.75 for 1.75 meters
# Your code here:


In [ ]:
# Exercise 4: Create three variables:
# - city (string with your city name)
# - temperature (float with current temperature)
# - is_sunny (we'll use a string "yes" or "no" for now)
# Print all three
# Your code here:


In [ ]:
# Exercise 5: Variable reassignment
# Create a variable 'count' with value 1
# Print it
# Change it to 2
# Print it again
# Your code here:


### 🎮 Challenge: Using Variables Together

In [8]:
# Create variables for your first name and last name
# Then print them together
# Example output: "John Doe"

first_name = "John"
last_name = "Doe"
print(first_name, last_name)  # Notice: print can take multiple values!

# Now you try with your own name:


John Doe


### 🚫 Common Mistakes to Avoid

In [10]:
# ❌ MISTAKE 1: Trying to use a variable before creating it
# Uncomment the line below to see the error:
# print(unknown_variable)

# ✅ CORRECT: Create the variable first
unknown_variable = "Now it exists!"
print(unknown_variable)

Now it exists!


In [ ]:
# ❌ MISTAKE 2: Putting the variable on the right side
# This is wrong: 42 = x

# ✅ CORRECT: Variable name on the left, value on the right
x = 42
print(x)

In [ ]:
# ⚠️ NOTE: Variable names are case-sensitive
message = "lowercase"
Message = "Uppercase M"

print(message)  # Different from Message!
print(Message)

### 📏 Variable Naming Rules

**Rules you MUST follow:**
- ✅ Can contain letters, numbers, and underscores
- ✅ Must start with a letter or underscore (not a number)
- ✅ Case-sensitive (`name` ≠ `Name`)
- ❌ Cannot use Python keywords (like `print`, `if`, `for`)
- ❌ Cannot contain spaces

**Good variable names:**
```python
name
user_age
total_count
firstName
_private_var
```

**Bad variable names:**
```python
2name         # Can't start with number
user-age      # Can't use hyphens
total count   # Can't have spaces
for           # Can't use Python keywords
```

In [ ]:
# Practice: Try creating variables with good names
# Your code here:


---
### 📌 Key Takeaways

- ✅ **Variables** are names that refer to values
- ✅ Use `=` for **assignment** (variable = value)
- ✅ Assignment creates new variables
- ✅ Variables can store different **types**: strings, integers, floats
- ✅ You can **reassign** variables to new values
- ✅ Variable names must follow specific **rules**
- ✅ Variables make programs **powerful** and **flexible**

**Remember:** Think of variables as labeled containers for your data!

### 🤔 Reflection Questions

1. Why are variables called "variables"? (Hint: they can vary/change)
2. What's the difference between `x = 5` and `5 = x`?
3. Can you create a variable called `print`? Should you?
4. What happens if you create `age = 25` and then `age = 26`?

**Try experimenting below to find the answers!**

In [ ]:
# Experiment area


### 💭 My Notes

*Write down your thoughts, questions, or important points:*

- 
- 
- 

---
### 🎯 Next Steps

Now that you understand variables, you're ready to learn about:
- Variable names and keywords
- Expressions and operators
- How to use variables in calculations

**Keep going! You're building a strong foundation!** 🚀